src/parser/statement.js

```diff
@@ -94,6 +94,7 @@ export default class StatementParser extends ExpressionParser {
       case tt._for:
         return this.parseForStatement(node);
       case tt._function:
+        if (this.lookahead().type === tt.dot) break;
         if (!declaration) this.unexpected();
```

Comment:
Maybe you can add a test which checks the order of these two conditions.

e.g.
```js
function* gen() {
  if (true) function.sent;
}
```